packages/database/src/modelClasses/AncestorDescendantRelation.js

```diff
@@ -3,25 +3,59 @@
  * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
  */
 
-import { reduceToDictionary } from '@tupaia/utils';
+import { reduceToDictionary, reduceToArrayDictionary } from '@tupaia/utils';
 import { DatabaseModel } from '../DatabaseModel';
 import { DatabaseType } from '../DatabaseType';
 import { TYPES } from '../types';
 
-class AncestorDescendantRelationType extends DatabaseType {
+export class AncestorDescendantRelationType extends DatabaseType {
   static databaseType = TYPES.ANCESTOR_DESCENDANT_RELATION;
+
+  static joins = [
+    {
+      joinWith: TYPES.ENTITY,
+      joinAs: 'descendant',
+      joinCondition: ['descendant_id', 'descendant.id'],
+      fields: { code: 'descendant_code' },
+    },
+    {
+      joinWith: TYPES.ENTITY,
+      joinAs: 'ancestor',
+      joinCondition: ['ancestor_id', 'ancestor.id'],
+      fields: { code: 'ancestor_code' },
+    },
+  ];
 }
 
 export class AncestorDescendantRelationModel extends DatabaseModel {
   get DatabaseTypeClass() {
     return AncestorDescendantRelationType;
   }
 
-  async getChildIdToParentId(hierarchyId) {
-    const relationRecords = await this.find({
+  async getImmediateRelations(hierarchyId) {
+    return this.find({
       entity_hierarchy_id: hierarchyId,
       generational_distance: 1,
     });
+  }
+
+  async getChildIdToParentId(hierarchyId) {
+    const relationRecords = await this.getImmediateRelations(hierarchyId);
     return reduceToDictionary(relationRecords, 'descendant_id', 'ancestor_id');
   }
+
+  async getChildCodeToParentCode(hierarchyId) {
+    const relationRecords = await this.getImmediateRelations(hierarchyId);
+    return reduceToDictionary(relationRecords, 'descendant_code', 'ancestor_code');
+  }
+
+  async getParentIdToChildIds(hierarchyId) {
```

Comment:
There were a few iterations of work getting this going. Initially I was pulling just the `ids` in and then mapping them to codes based on the entity objects returned from the `getDesendants()` fetch. This is how we do it currently in `/organisationUnit` in `web-config-server`. However the drawback there is that:
a) You can't show the parent/child of entities not in that fetch
b) The logic is just a bit more awkward and convoluted

So I switched to doing the join in the db layer and fetching the codes directly, however thought it might be best to leave a few of these functions around that aren't doing any harm?

Reply:
Yeah that's fair enough. I'm tempted to remove this one, just because it's one less bit of clutter, but it's honestly no big deal either way